doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,248 @@
+# Initial cluster SQL API
+
+## Summary
+
+This document proposes the initial SQL API changes to introduce
+[clusters](../platform/ux.md#cluster) to Materialize.
+
+## Design
+
+### Cluster SQL syntax and semantics
+
+Four new SQL statements will be introduced with the following grammar to
+create, alter, drop, and list clusters, respectively:
+
+```
+create_cluster_stmt ::=
+  CREATE CLUSTER [IF NOT EXISTS] <name> [[WITH] <option> [, <option>...]]
+
+alter_cluster_stmt ::=
+  ALTER CLUSTER <name> [[SET] <option> [, <option>...]]
+
+drop_cluster_stmt ::=
+  DROP CLUSTER [IF EXISTS] <name> [CASCADE]
+
+show_cluster_stmt: SHOW CLUSTERS [{ LIKE 'pattern' | WHERE <expr> }]
+
+<option> ::=
+      VIRTUAL
+    | SIZE '<size>'
+
+<name> ::= <unqualified-identifier>
+```
+
+**All of these statements will be available in experimental mode only.**
+
+The name of a cluster is an unqualified identifier. Clusters are scoped to
+a catalog, and so do not belong in the usual namespace of databases and schemas.
+For a similarly scoped object in PostgreSQL, see [publications].
+
+The `CREATE CLUSTER` statement creates a new cluster. It returns an error if
+a cluster with the specified name already exists unless the `IF NOT EXISTS`
+clause is present.
+
+The `ALTER CLUSTER` statement changes properties of an existing cluster.
+
+The `DROP CLUSTER` statement drops an existing cluster. It returns an error
+if a cluster with the specified name does not exist unless the `IF EXIST`
+clause is present. Dropping a non-empty cluster (i.e., a cluster with at least
+one index) is not permitted unless `CASCADE` is present.
+
+THe `SHOW CLUSTERS` statement lists the clusters in the system, optionally
+filtered by the provided `LIKE` pattern or `WHERE` expression, which work
+analogously to the same clauses in the `SHOW DATABASES` statement.
+
+`CREATE CLUSTER` and `ALTER CLUSTER` allow specifying the same set of cluster
+options. Initially, there are only two options:
+
+  * `VIRTUAL` requests a virtual cluster. A virtual cluster may or may not
+    correspond to actual physical resources.
+
+    We suspect that the concept of a virtual cluster is primarily useful as a
+    way to exercise the cluster logic before we can allocate true clusters, and
+    it is likely that the `VIRTUAL` option is never stabilized. But they may
+    also prove to be a useful user-facing option, in which case we can consider
+    stabilization.
```

Comment:
I find the `VIRTUAL` reasoning somewhat unclear. Can you add a note what it refers to? To me it's unclear whether this is the in-process cluster that materialized offers, or a cluster that accepts sql statements but .. doesn't do anything?

Reply:
It's intentionally vague, is the thing! @frankmcsherry's take is that putting indexes in the same cluster guarantees that they can see one another, but the inverse is not necessarily true. Putting indexes in _different_ clusters does not guarantee that they will _not_ share physical resources with indexes in another cluster. It's up to COMPUTE to decide what to do with these separate cluster IDs in the short term when we only support a single process, and from the perspective of ADAPTER we're being intentionally vague about what COMPUTE is or is not doing.

It's all kind of a hack, IMO, to get us exercising `CREATE CLUSTER` and `DROP CLUSTER` before we have support for real clusters. I suspect we'll delete the concept of creating/deleting clusters without ever stabilizing them, and instead single-process mode will have one undroppable built-in virtual cluster named `default`. But, maybe we'll decide that allowing suers to create/drop virtual clusters is actually a useful thing, and is worth explaining and stabilizing.